src/libraries/System.Private.Xml/src/System/Xml/Schema/FacetChecker.cs

```diff
@@ -330,6 +330,13 @@ internal void CompileFractionDigitsFacet(XmlSchemaFacet facet)
                 CheckDupFlag(facet, RestrictionFlags.FractionDigits, SR.Sch_DupFractionDigitsFacet);
                 _derivedRestriction.FractionDigits = XmlBaseConverter.DecimalToInt32((decimal)ParseFacetValue(_nonNegativeInt, facet, SR.Sch_FractionDigitsFacetInvalid, null, null));
 
+                if ((_baseFixedFlags & RestrictionFlags.FractionDigits) != 0)
```

Comment:
I think before this check better to check `if ((_derivedRestriction.FractionDigits != 0) && (_datatype.TypeCode != XmlTypeCode.Decimal))` and throw corresponding exception, please move